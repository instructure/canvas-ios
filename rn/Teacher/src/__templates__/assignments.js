//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* @flow */

import template, { type Template } from '../utils/template'

export const assignment: Template<Assignment> = template({
  id: '123456789',
  course_id: '987654321',
  name: 'learn to write code',
  description: 'a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
  created_at: '2017-03-17T19:15:25Z',
  updated_at: '2017-03-17T19:15:25Z',
  points_possible: 20,
  grading_type: 'points',
  assignment_group_id: '1376',
  peer_reviews: false,
  automatic_peer_reviews: false,
  position: 1,
  submission_types:
  [ 'online_text_entry',
    'online_url',
    'media_recording',
    'online_upload' ],
  needs_grading_count: 0,
  html_url: 'http://hostname.tld/whatever',
  overrides: [],
  grade_group_students_individually: false,
  group_category_id: null,
  has_overrides: false,
  only_visible_to_overrides: false,
  published: true,
  unpublishable: false,
  rubric_assessment: {},
  rubric_settings: null,
  rubric: null,
  anonymize_students: false,
})

export const assignmentGroup: Template<AssignmentGroup> = template({
  id: '1',
  name: 'Learn React Native',
  position: 1,
  assignments: [assignment({ assignment_group_id: '1' })],
  group_weight: 1,
  integration_data: {},
  sis_source_id: '',
})

export const assignmentDueDate: Template<AssignmentDate> = template({
  title: 'due',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})

export const assignmentOverride: Template<AssignmentOverride> = template({
  id: '1',
  assignment_id: '1',
  student_ids: [],
  group_id: '1',
  course_section_id: '1',
  title: 'override',
  all_day: false,
  all_date_date: '',
  due_at: '2037-06-01T05:59:00Z',
  unlock_at: '2017-03-17T06:00:00Z',
  lock_at: '2037-06-01T05:59:59Z',
})
