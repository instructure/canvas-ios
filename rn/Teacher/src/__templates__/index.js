//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

export * from '../redux/__templates__/app-state'
export * from './account-notification'
export * from './addressBook'
export * from './api'
export * from './assignments'
export * from './attachment'
export * from './conversations'
export * from './course'
export * from './discussion'
export * from './enrollments'
export * from './error'
export * from './external-tool'
export * from './file'
export * from './grading-periods'
export * from './group'
export * from './helm'
export * from './launch-definitions'
export * from './mediaComment'
export * from './quiz'
export * from './quizSubmission'
export * from './role'
export * from './rubric'
export * from './section'
export * from './session'
export * from './submissions'
export * from './tab'
export * from './users'
export * from './help-links'
export * from './dashboard'
export * from '../modules/assignee-picker/__template__/Assignee'
