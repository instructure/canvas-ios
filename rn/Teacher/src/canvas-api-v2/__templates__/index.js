// @flow
export * from './submission'
export * from './user'
export * from './assignment'
export * from './course'
export * from './group'
export * from './section'
export * from './groupSet'

// QUERIES
export * from '../queries/__templates__/SubmissionList'
export * from '../queries/__templates__/ContextCard'
