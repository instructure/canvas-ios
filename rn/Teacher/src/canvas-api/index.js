//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

export { default } from './apis/index'
export * from './apis/assignmentGroups'
export * from './apis/assignments'
export * from './apis/conversations'
export * from './apis/courses'
export * from './apis/enrollments'
export * from './apis/external-tools'
export * from './apis/groups'
export * from './apis/login'
export * from './apis/quizzes'
export * from './apis/submissions'
export * from './apis/users'
export * from './apis/media-comments'
export * from './apis/feature-flags'
export * from './apis/file-uploads'
export * from './apis/accounts'
export * from './apis/dashboard'

export * from './session'

export { default as httpClient, isAbort, httpCache } from './httpClient'
