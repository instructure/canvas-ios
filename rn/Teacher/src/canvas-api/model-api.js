//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
export * from './fetch-props-for'
export * from './model'
export { httpCache } from './httpClient'

import httpClient, { httpCache } from './httpClient'
import parseLink from './utils/parse-link-header'
import {
  CourseModel,
  PageModel,
  ToDoModel,
} from './model'

type ApiPolicy = 'cache-only' | 'cache-and-network' | 'network-only'
type ApiOptions = {|
  policy: ApiPolicy,
  onStart?: ApiPromise<any> => void,
  onComplete?: ApiPromise<any> => void,
  onAnyComplete?: ApiPromise<any> => void,
  onError?: Error => void,
|}
type Paginated<T> = {|
  list: T,
  next: ?string,
  getNextPage: ?(() => ApiPromise<Paginated<T>>),
|}

const emptyArray = []
const emptyPaginated: Paginated<any> = {
  list: emptyArray,
  next: null,
  getNextPage: null,
}

export class API {
  options: ApiOptions

  constructor (options: ApiOptions) {
    this.options = options
  }

  cleanup () {
    this.options.onStart =
    this.options.onComplete =
    this.options.onAnyComplete =
    this.options.onError = undefined
  }

  get (url: string, config: ApiConfig = {}) {
    const { policy } = this.options
    const cached = httpCache.get(url, config)
    if (
      policy === 'network-only' ||
      (policy === 'cache-and-network' && cached.expiresAt < Date.now())
    ) {
      this.request(httpClient.get(url, config))
    }
    return cached.value || null
  }

  paginate<T> (url: string, config: ApiConfig = {}): Paginated<T> {
    const transform = config.transform
    config.transform = (value: any, response: ApiResponse<any>) => {
      const list = transform ? transform(value, response) : value
      const { current, first, next } = parseLink(response.headers.link) || {}
      const getNextPage = next && (() => {
        this.request(httpClient.get(next, config))
      })
      if (getNextPage && config.params && config.params.per_page >= 100) {
        getNextPage() // automatically request all pages
      }
      if (current && current !== first) {
        const base = httpCache.get(url, config).value
        if (base) {
          base.list = base.list.concat(list)
          base.next = next
          base.getNextPage = getNextPage
        }
        return null
      }
      return {
        list,
        next,
        getNextPage,
      }
    }

    const cached = httpCache.get(url, config)
    const { policy } = this.options
    if (
      policy === 'network-only' ||
      (policy === 'cache-and-network' && cached.expiresAt < Date.now())
    ) {
      this.request(httpClient.get(url, config))
    }
    // getNextPage won't survive serialization
    const value = cached.value
    if (value && value.next && !value.getNextPage && policy !== 'cache-only') {
      value.getNextPage = () => this.request(httpClient.get(value.next, config))
    }
    return value || emptyPaginated
  }

  post (url: string, data: *, config: ApiConfig = {}) {
    return this.request(httpClient.post(url, data, config))
  }

  put (url: string, data: *, config: ApiConfig = {}) {
    return this.request(httpClient.put(url, data, config))
  }

  delete (url: string, config: ApiConfig = {}) {
    return this.request(httpClient.delete(url, config))
  }

  request (promise: ApiPromise<any>) {
    const { onStart, onComplete, onError } = this.options
    onStart && onStart(promise)
    promise
      .catch(error => { onError && onError(error) })
      .then(() => { onComplete && onComplete(promise) })
    return promise
  }

  getCourseColor (courseID: string): string {
    const json: ?Object = this.get('users/self/colors')
    return json && json.custom_colors[`course_${courseID}`] || '#aaa'
  }

  getCourse (courseID: string): ?CourseModel {
    return this.get(`courses/${courseID}`, {
      params: {
        include: [ 'permissions', 'term', 'favorites', 'course_image', 'sections' ],
      },
      transform: (course: Course) => new CourseModel(course),
    })
  }

  getPages (context: CanvasContext, contextID: string): Paginated<PageModel[]> {
    return this.paginate(`${context}/${contextID}/pages`, {
      params: {
        per_page: 100,
      },
      transform: (pages: Page[]) => pages.map(page => new PageModel(page)),
    })
  }

  createPage (context: CanvasContext, contextID: string, parameters: PageParameters): ApiPromise<PageModel> {
    return this.post(`${context}/${contextID}/pages`, { wiki_page: parameters }, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  getPage (context: CanvasContext, contextID: string, url: string): ?PageModel {
    return this.get(`${context}/${contextID}/pages/${url}`, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  getFrontPage (contextID: string): ?PageModel {
    return this.get(`courses/${contextID}/front_page`, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  updatePage (context: CanvasContext, contextID: string, url: string, parameters: PageParameters): ApiPromise<PageModel> {
    return this.put(`${context}/${contextID}/pages/${url}`, { wiki_page: parameters }, {
      transform: (page: Page) => new PageModel(page),
    })
  }

  deletePage (context: CanvasContext, contextID: string, url: string): ApiPromise<null> {
    return this.delete(`${context}/${contextID}/pages/${url}`)
  }

  getToDos (): Paginated<ToDoModel[]> {
    return this.paginate('users/self/todo', {
      transform: (todos: ToDoItem[]) => todos.map(todo => new ToDoModel(todo)),
    })
  }
}
