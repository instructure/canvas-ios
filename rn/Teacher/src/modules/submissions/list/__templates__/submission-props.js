//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import template, { type Template } from '../../../../utils/template'
import type { SubmissionDataProps } from '../submission-prop-types'

const templates = {
  ...require('../../../../__templates__/submissions'),
}

const submissionWithHistory = templates.submissionHistory()

// $FlowFixMe
export const submissionProps: Template<SubmissionDataProps> = template({
  grade: 'B-',
  name: submissionWithHistory.user.name,
  status: 'submitted',
  userID: submissionWithHistory.user.id,
  avatarURL: submissionWithHistory.user.avatar_url,
  sectionID: '1',
})
