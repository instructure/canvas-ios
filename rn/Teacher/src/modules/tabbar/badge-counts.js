//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
import { NativeModules } from 'react-native'
import canvas from '../../canvas-api'
import { isTeacher } from '../app'

export let interval: ?any

const TabBarBadgeCounts = NativeModules.TabBarBadgeCounts

export async function updateBadgeCounts () {
  try {
    const unread = canvas.getUnreadConversationsCount().then(({ data: unread }) =>
      TabBarBadgeCounts.updateUnreadMessageCount(+unread.unread_count || 0)
    )
    if (isTeacher()) {
      await canvas.getToDoCount().then(({ data: todo }) =>
        TabBarBadgeCounts.updateTodoListCount(+todo.needs_grading_count)
      )
    }
    await unread
  } catch (e) {}
}

export async function beginUpdatingBadgeCounts () {
  stopUpdatingBadgeCounts()
  interval = setInterval(updateBadgeCounts, 2 * 60 * 1000) // every 2 minutes
  await updateBadgeCounts()
}

export function stopUpdatingBadgeCounts (): void {
  if (!interval) {
    return
  }
  clearInterval(interval)
  interval = undefined
}
