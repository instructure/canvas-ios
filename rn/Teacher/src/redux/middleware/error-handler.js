//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import { Alert } from 'react-native'
import type { MiddlewareAPI } from 'redux'
import i18n from 'format-message'
import loginVerify from '../../common/login-verify'

export function alertError (error: any, alertTitle?: string, callback?: Function): void {
  const title = alertTitle || defaultErrorTitle()
  const message = parseErrorMessage(error)
  let buttons = [{ text: i18n('Dismiss'), onPress: () => { if (callback) callback() } }]
  Alert.alert(title, message, buttons)

  if (error instanceof Error) {
    console.warn(`Error: ${error.message}.  Stack:\n${error.stack}`)
  } else {
    console.warn(`Error: ${error}`)
  }
}

export function defaultErrorTitle (): string {
  return i18n('Unexpected Error')
}

export function defaultErrorMessage (): string {
  return i18n('There was an unexpected error. Please try again.')
}

let showingGlobalErrorAlert = false
const showGlobalErrorAlertIfNecessary = async (error: any) => {
  if (showingGlobalErrorAlert) return
  showingGlobalErrorAlert = true
  if (error.response != null) {
    error = error.response
  }

  let alertTitle

  alertError(error, alertTitle, () => {
    resetGlobalErrorAlert()
  })
}

export function resetGlobalErrorAlert () {
  showingGlobalErrorAlert = false
}

const errorHandlerMiddleware: MiddlewareAPI = () => {
  return next => action => {
    if (action.error) {
      let error = action.payload.error
      if (error && error.response && error.response.status === 401) {
        loginVerify().then((invalid) => {
          if (!invalid) {
            if (action.payload.handlesError) return next(action)
            showGlobalErrorAlertIfNecessary(error)
          }
          next(action)
        })
        return action // Bail out of this because of the promise. Next will be called after the promise has finished
      } else {
        if (action.payload.handlesError) return next(action)
        showGlobalErrorAlertIfNecessary(error)
      }
    }
    return next(action)
  }
}

export default errorHandlerMiddleware

export function parseErrorMessage (error: any): string {
  if (error && error.response) error = error.response

  if (typeof error === 'string') {
    return error
  }

  if (error && error.data && error.data.message) {
    return error.data.message
  } else if (error && error.data && error.data.errors && error.data.errors.length > 0) {
    if (typeof error.data.errors === 'string') {
      return error.data.errors
    }
    return error.data.errors
      .map(error => error.message)
      .map(message => message.replace(/\.+$/, ''))
      .join('. ')
  } else if (error && error.data && error.data.errors instanceof Object && Object.keys(error.data.errors).length > 0) {
    let data = error.data.errors
    let result = Object.keys(data)
      .map((key, index) => {
        let err = data[key]
        if (Array.isArray(err) && err.length > 0) err = err[0]
        if (typeof err === 'string') return err
        return err && err.message || ''
      })
      .join('. ')

    if (result.trim()) {
      return result
    }
  }

  if (error instanceof Error) {
    return error.message
  }

  return defaultErrorMessage()
}
