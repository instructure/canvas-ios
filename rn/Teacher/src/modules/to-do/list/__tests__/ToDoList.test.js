//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import { shallow } from 'enzyme'
import React from 'react'
import { API, httpCache } from '../../../../canvas-api/model-api'
import * as template from '../../../../__templates__'
import { alertError } from '../../../../redux/middleware/error-handler'
import { updateBadgeCounts } from '../../../tabbar/badge-counts'
import Connected, { ToDoList } from '../ToDoList'

jest.mock('../../../../redux/middleware/error-handler', () => {
  return { alertError: jest.fn() }
})

jest.mock('../../../tabbar/badge-counts', () => ({
  updateBadgeCounts: jest.fn(),
}))

describe('ToDoList', () => {
  let props
  beforeEach(() => {
    props = {
      navigator: template.navigator(),
      list: [template.toDoModel()],
      getNextPage: jest.fn(),
      api: new API({ policy: 'cache-only' }),
      isLoading: false,
      loadError: null,
      refresh: jest.fn(),
    }
  })

  it('gets to dos from the model api', () => {
    const list = [ template.toDoModel() ]
    const next = 'users/self/todo?page=2'
    const getNextPage = jest.fn()
    httpCache.handle('GET', 'users/self/todo', { list, next, getNextPage })
    const tree = shallow(<Connected />)
    expect(tree.find(ToDoList).props()).toMatchObject({
      list,
      next,
      getNextPage,
    })
  })

  it('renders items', () => {
    const tree = shallow(<ToDoList {...props} />)
    expect(tree).toMatchSnapshot()
  })

  it('alerts when there is a load error', () => {
    const tree = shallow(<ToDoList {...props} />)
    tree.setProps({ loadError: null })
    expect(alertError).not.toHaveBeenCalled()
    const loadError = new Error()
    tree.setProps({ loadError })
    expect(alertError).toHaveBeenCalledWith(loadError)
  })

  it('updates tab bar todo count on refresh', () => {
    const tree = shallow(<ToDoList {...props} />)
    tree.find('FlatList').simulate('Refresh')
    expect(updateBadgeCounts).toHaveBeenCalled()
    expect(props.refresh).toHaveBeenCalled()
  })

  it('gets next page immediately if there are less than 10', () => {
    const tree = shallow(<ToDoList {...props} />)
    const getNextPage = jest.fn()
    tree.setProps({
      list: [
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '1' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '2' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '3' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '4' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '5' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '6' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '7' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '8' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '9' }),
        template.toDoModel({ needsGradingCount: 0, htmlUrl: '10' }),
      ],
      getNextPage,
    })
    expect(getNextPage).toHaveBeenCalled()
  })

  it('gets next page on end reached', async () => {
    const tree = shallow(<ToDoList {...props} />)
    tree.find('FlatList').simulate('EndReached')
    expect(props.getNextPage).toHaveBeenCalled()
  })

  it('shows Speed Grader on assignment press', () => {
    props.navigator = template.navigator({ show: jest.fn() })
    props.list = [template.toDoModel({
      courseID: '222',
      assignment: template.assignment({ id: '1' }),
    })]
    const tree = shallow(<ToDoList {...props} />)
    tree.find('FlatList').dive().find('ToDoListItem')
      .simulate('Press', props.list[0])
    expect(props.navigator.show).toHaveBeenCalledWith(
      '/courses/222/assignments/1/submissions?filter=needs_grading',
      { modal: true, modalPresentationStyle: 'fullscreen', embedInNavigationController: false },
    )
  })

  it('shows Speed Grader on quiz press', () => {
    props.navigator = template.navigator({ show: jest.fn() })
    props.list = [template.toDoModel({
      courseID: '222',
      assignment: null,
      quiz: template.quiz({ id: '2', assignment_id: '1' }),
    })]
    const tree = shallow(<ToDoList {...props} />)
    tree.find('FlatList').dive().find('ToDoListItem')
      .simulate('Press', props.list[0])
    expect(props.navigator.show).toHaveBeenCalledWith(
      '/courses/222/assignments/1/submissions?filter=needs_grading',
      { modal: true, modalPresentationStyle: 'fullscreen', embedInNavigationController: false },
    )
  })
})
