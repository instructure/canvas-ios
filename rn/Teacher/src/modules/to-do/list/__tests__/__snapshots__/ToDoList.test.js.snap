// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList renders items 1`] = `
<Screen
  customPageViewPath="/"
  drawUnderNavBar={true}
  navBarButtonColor="#374A59"
  navBarColor="#374A59"
  navBarImage={
    Object {
      "testUri": "../../../src/images/canvas-logo.png",
    }
  }
  navBarStyle="light"
  showDismissButton={true}
  statusBarStyle="light"
>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <FlatList
      ItemSeparatorComponent={[Function]}
      ListEmptyComponent={
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "white",
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              },
              Object {
                "height": 0,
              },
            ]
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/relax.png",
              }
            }
            style={
              Object {
                "marginBottom": 24,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "#8B969E",
                "fontSize": 18,
              }
            }
          >
            Nothing more to do.
          </Text>
          <Text
            style={
              Object {
                "color": "#8B969E",
                "fontSize": 18,
              }
            }
          >
            Enjoy your day!
          </Text>
        </View>
      }
      data={
        Array [
          ToDoModel {
            "assignment": Object {
              "anonymize_students": false,
              "assignment_group_id": "1376",
              "automatic_peer_reviews": false,
              "course_id": "987654321",
              "created_at": "2017-03-17T19:15:25Z",
              "description": "a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.",
              "due_at": "2037-06-01T05:59:00Z",
              "grade_group_students_individually": false,
              "grading_type": "points",
              "group_category_id": null,
              "has_overrides": false,
              "html_url": "http://hostname.tld/whatever",
              "id": "123456789",
              "lock_at": "2037-06-01T05:59:59Z",
              "name": "learn to write code",
              "needs_grading_count": 0,
              "only_visible_to_overrides": false,
              "overrides": Array [],
              "peer_reviews": false,
              "points_possible": 20,
              "position": 1,
              "published": true,
              "rubric": null,
              "rubric_assessment": Object {},
              "rubric_settings": null,
              "submission_types": Array [
                "online_text_entry",
                "online_url",
                "media_recording",
                "online_upload",
              ],
              "unlock_at": "2017-03-17T06:00:00Z",
              "unpublishable": false,
              "updated_at": "2017-03-17T19:15:25Z",
            },
            "contextType": "course",
            "courseID": "1",
            "groupID": null,
            "htmlUrl": "https://api.instructure.com/api/v1/path/to/html_url",
            "ignorePermanentlyUrl": "https://api.instructure.com/api/v1/path/to/ignore_permanently",
            "ignoreUrl": "https://api.instructure.com/api/v1/path/to/ignore",
            "needsGradingCount": 5,
            "quiz": undefined,
            "raw": Object {
              "assignment": Object {
                "anonymize_students": false,
                "assignment_group_id": "1376",
                "automatic_peer_reviews": false,
                "course_id": "987654321",
                "created_at": "2017-03-17T19:15:25Z",
                "description": "a course that can take any beginner to professional in a matter of microseconds.  Harness the power of your favorite languages like PHP and Cold Fusion.",
                "due_at": "2037-06-01T05:59:00Z",
                "grade_group_students_individually": false,
                "grading_type": "points",
                "group_category_id": null,
                "has_overrides": false,
                "html_url": "http://hostname.tld/whatever",
                "id": "123456789",
                "lock_at": "2037-06-01T05:59:59Z",
                "name": "learn to write code",
                "needs_grading_count": 0,
                "only_visible_to_overrides": false,
                "overrides": Array [],
                "peer_reviews": false,
                "points_possible": 20,
                "position": 1,
                "published": true,
                "rubric": null,
                "rubric_assessment": Object {},
                "rubric_settings": null,
                "submission_types": Array [
                  "online_text_entry",
                  "online_url",
                  "media_recording",
                  "online_upload",
                ],
                "unlock_at": "2017-03-17T06:00:00Z",
                "unpublishable": false,
                "updated_at": "2017-03-17T19:15:25Z",
              },
              "context_type": "course",
              "course_id": "1",
              "group_id": null,
              "html_url": "https://api.instructure.com/api/v1/path/to/html_url",
              "ignore": "https://api.instructure.com/api/v1/path/to/ignore",
              "ignore_permanently": "https://api.instructure.com/api/v1/path/to/ignore_permanently",
              "needs_grading_count": 5,
              "type": "grading",
            },
            "type": "grading",
          },
        ]
      }
      keyExtractor={[Function]}
      onEndReached={[MockFunction]}
      onEndReachedThreshold={0.5}
      onRefresh={[Function]}
      refreshing={false}
      renderItem={[Function]}
      testID="to-do-list.list"
    />
  </View>
</Screen>
`;
